src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -888,6 +905,79 @@ internal void ExecuteSubmission(BuildSubmission submission, bool allowMainThread
             }
         }
 
+        /// <summary>
+        /// This method adds the graph build request in the specified submission to the set of requests being handled by the scheduler.
+        /// </summary>
+        internal void ExecuteGraphSubmission(BuildSubmission submission)
+        {
+            lock (_syncLock)
+            {
+                VerifyStateInternal(BuildManagerState.Building);
+
+                try
+                {
+                    // Create the build request
+                    submission.BuildRequest = new BuildRequest(
+                        submission.SubmissionId,
+                        BackEnd.BuildRequest.InvalidNodeRequestId,
+                        GetNewConfigurationId(), // TODO: Should we create a real BuildRequestConfiguration for each graph request?
+                        submission.BuildRequestData.TargetNames,
+                        submission.BuildRequestData.HostServices,
+                        BuildEventContext.Invalid,
+                        null,
+                        submission.BuildRequestData.Flags,
+                        submission.BuildRequestData.RequestedProjectState);
+
+                    if (_shuttingDown)
+                    {
+                        // We were already canceled!
+                        BuildResult result = new BuildResult(submission.BuildRequest, new BuildAbortedException());
+                        submission.CompleteResults(result);
+                        submission.CompleteLogging(true);
+                        CheckSubmissionCompletenessAndRemove(submission);
+                        return;
+                    }
+
+                    if (_graphSchedulingCancellationSource == null)
+                    {
+                        _graphSchedulingCancellationSource = new CancellationTokenSource();
```

Comment:
I think it'd be a little cleaner to do something like

```c#
CancellationToken schedulingCancellationToken = _graphSchedulingCancellationSource?.Token ?? CancellationToken.None
```

Reply:
Well, that's semantically different. Effectively I'm just wanting to use 1 cancellation source for the BuildManager across sessions (ignoring EndBuilds + new BeginBuilds), so this is basically just lazily creating it. I don't want to use `CancellationToken.None` (which would possibly leak the thread), but instead the intent here is to use the existing cancellation source if there is one, and if there isn't one just create one and use that.

Reply:
Oh, I see--there is a codepath to call cancel on a token created here. Sounds good, but . . . comments :)